### DISCUSSION TITLE
DataLinq Rest Query - Möglichkeit von select distinct

### DISCUSSION THREAD

**Discussion by faulgurke** [e-netze/webgis-community#40] [Q&A]
Gibt es für die DataLinq Rest Query (URL/gebaeude?gebaeudeart={{GebaeudeArt}}) die Möglichkeit ein select distinct zu formulieren?
Wenn ja, wie macht man das?

**Comment by gis-eni** [comment]
Im DataLinq wird das SELECT Statement definiert. Beim Abholen werden die Daten über dieses Statement abgeholt, nicht mehr und nicht weniger. Das ist auch gut so. Das Statement sollte nicht über die Url geändert werden können, außer den Parametern, die du im Statement definierst. Alles andere wäre Sicherheitslücke (SQL Injection)

Ja, du kannst ein Distinct machen. Bau es in dein SELECT Statement ein, bzw. mach eine eigene Abfrage für Abfrage dafür.
Du kannst unterhalb eines Endpoints beliebig viele *Queries* anlagen (gebauede, gebaude-distinct, ...)

**Reply by faulgurke** [reply]
Ich meine keine SQL-Abfrage, sondern eine REST-Abfrage. Also eine Query in der Form:
URL/gebaeude?gebaeudeart={{GebaeudeArt}}
#if dachfarbe
        &farbe = @dachfarbe
#endif
#if _orderby
        &farbe = @dachfarbe
#endif
#if _orderby
    &_orderby={{_orderby}}
#endif

Kann man da irgendwo ein Select distinct einbauen?
Gibt es die Möglichkeit, das Select distinct im Report zu formulieren? (z.B für @DLH.FilterView)

**Comment by gis-eni** [comment]
Das hängt dann natürlich von deiner REST Schnittstelle ab. Auf was fragst du ab?

**Reply by faulgurke** [reply]
Das ist ein Service mit den stehenden Gewässern Vorarlbergs. Die Attribute sind Gemeinde, Typ (z.b. Stausee), Name, Fläche.
Es soll eine Auswahlliste erstellt werden, in welcher jeder **Typ** nur einmal vorkommt (in SQL wäre das ein Select Distinct).
Die daraus resultierende Query wird dann in einer DLH.FilterView eingebaut.

**Comment by gis-eni** [comment]
Und kann dieses Service ein Distinct? Wer hat dieses Service gemacht?

**Reply by faulgurke** [reply]
gView 5.22.4601. Ob das Service ein Select Distinct unterstützt, weiß ich nicht.

**Comment by gis-eni** [comment]
gView unterstützt die REST Schnittstelle von ArcGIS Server. Dort gibt es auch ein Distinct (bin mir aber nicht sicher, ob das dieser Version schon dabei war)
Du kannst in der ESRI Doku nachlesen, wie das geht.

Wie bindest du jetzt einen gView REST Query in Datalinq ein?

**Reply by faulgurke** [reply]
Wie wird die Query eingebunden:
gewaesser_gv/queries/seen?geometry=none
#if klasse
    &klasse={{klasse}}
#endif
Der Versuch entsprechend der ESRI Doku war nicht erfolgreich.
Doku:
https://developers.arcgis.com/rest/services-reference/enterprise/query-feature-service-layer-.htm
Query:
gewaesser_gv/queries/seen?geometry=none&outFields=klasse&returnDistinctValues=true
(die Query gewaesser_gv/queries/seen?geometry=none funktioniert problemlos)

**Comment by gis-eni** [comment]
Ah, schön langsam kommen ich dahinter, was du willst ;)

Du holst dir über DataLinq die Daten über die WebGIS REST API (wenn ich dieses Url-Fragement richtig interpretiere)
Da müsste natürlich die Überschrift "DISTINCT bei WebGIS REST Queries möglich/ermöglichen" heißen.

Fazit:
* Datalinq gibt die Daten so weiter, wie in der Abfrage efiniert => kein nachträgliches DISTINCT, bleibt so
* gView kann theoretisch ein DISTINCT über die REST Schnittstelle
* Dazwischen sitzt die WebGIS API, die eine Abfrage auf gView macht. Die Query-Schnittstelle, kann aber nur das, was man in WebGIS braucht, darum wurde hier keine DISTINCT implementiert.

Warum einfach, wenns auch kompliziert geht ;)

Die Antwort ist also: In dieser Konstellation ist das derzeit nicht möglich. Ob ein DISTINCT über die die WebGIS REST Queries implementiert werden kann und auch sinnvoll ist, muss ich erst prüfen.

Workarounds:
* Mit DataLinq direkt die Datenbank abfragen und im Statement das DISTINCT einbauen
* Direkt gView REST Schnittstelle abfragen: WebGIS API Query liefert ein GeoJson, gView Query kann ebenfalls ein GeoJSON zurückgeben => Es könnte funktionieren, wenn das GeoJSON mit Distinct direkt von gView holt.

**Reply by faulgurke** [reply]
Und ich hab' verstanden, was du meinst.
GeoServices (REST): da gibt's alle Parameter, die in der Doku stehen. gview/geoservices/rest/services/wasser/g5_gewaesser/FeatureServer/[LayerID]/query?outFields=klasse,gemeinde&returnDistinctValues=true
Allerdings ist mir noch nicht klar, wie ich die URL aus der DataLinq Query formulieren muss, damit ich die gView REST Schnittstelle direkt ansprechen kann.
